display all the available details of a finding

```cpp
#include <string>

static char* f() {
    return std::string().c_str();
}
```

Running Cppcheck on the command-line will show the following:

```
input.cpp:4:31: error: Returning object that will be invalid when returning. [returnDanglingLifetime]
    return std::string().c_str();
                              ^
input.cpp:4:31: note: Pointer to container is created here.
    return std::string().c_str();
                              ^
input.cpp:4:31: note: Returning object that will be invalid when returning.
    return std::string().c_str();
                              ^
input.cpp:4:5: error: Dangerous usage of c_str(). The value returned by c_str() is invalid after this call. [stlcstr]
    return std::string().c_str();
    ^
input.cpp:3:0: style: The function 'f' is never used. [unusedFunction]

^
```

But CLion only shows the first message for `returnDanglingLifetime`:

![image](https://user-images.githubusercontent.com/242857/168270742-b5278681-bc97-446e-893b-265c62a24da2.png)

For some findings this information is not sufficient to actually understand why it is being reported so we should provide these additional details in the IDE.

--- a/README.md
+++ b/README.md
@@ -144,6 +144,7 @@ Deployment.
 - Display `Cppcheck Path` configuration errors as global inspection errors instead of using a (hard to spot) status bar message.
 - Display global inspection error and omit the option if the configured `MISRA Addon JSON` does not exist.
 - Made plugin verbose level configurable via settings.
+- Display all available details for findings in tooltip.
 
 ### 1.6.2 - 2022-01-25
 

--- a/resources/META-INF/plugin.xml
+++ b/resources/META-INF/plugin.xml
@@ -28,6 +28,7 @@
     - Display `Cppcheck Path` configuration errors as global inspection errors instead of using a (hard to spot) status bar message.
     - Display global inspection error and omit the option if the configured `MISRA Addon JSON` does not exist.
     - Made plugin verbose level configurable via settings.
+    - Display all available details for findings in tooltip.
     ]]>
   </change-notes>
 

--- a/resources/META-INF/plugin.xml
+++ b/resources/META-INF/plugin.xml
@@ -23,11 +23,11 @@
     ]]></description>
 
   <change-notes><![CDATA[
-    - Added `Show Cppcheck XML Output` action to show the latest XML output.
-    - Report execution errors as global inspection errors.
-    - Display `Cppcheck Path` configuration errors as global inspection errors instead of using a (hard to spot) status bar message.
-    - Display global inspection error and omit the option if the configured `MISRA Addon JSON` does not exist.
-    - Made plugin verbose level configurable via settings.
+    - Added `Show Cppcheck XML Output` action to show the latest XML output.<br/>
+    - Report execution errors as global inspection errors.<br/>
+    - Display `Cppcheck Path` configuration errors as global inspection errors instead of using a (hard to spot) status bar message.<br/>
+    - Display global inspection error and omit the option if the configured `MISRA Addon JSON` does not exist.<br/>
+    - Made plugin verbose level configurable via settings.<br/>
     - Display all available details for findings in tooltip.
     ]]>
   </change-notes>

--- a/src/com/github/johnthagen/cppcheck/CppCheckInspectionImpl.java
+++ b/src/com/github/johnthagen/cppcheck/CppCheckInspectionImpl.java
@@ -169,30 +169,29 @@ class CppCheckInspectionImpl {
             final Node inconclusiveNode = attributes.getNamedItem("inconclusive");
             final boolean inconclusive = inconclusiveNode != null && inconclusiveNode.getNodeValue().equals("true");
 
-            Location location = null;
+            final ArrayList<Location> locations = new ArrayList<>();
 
             // look for the first "location" child name
             final NodeList children = error.getChildNodes();
             for (int j = 0; j < children.getLength(); ++j) {
                 final Node child = children.item(j);
                 if (child.getNodeName().equals("location")) {
-                    location = new Location(child);
-                    break;
+                    locations.add(new Location(child));
                 }
             }
 
             // ignore entries without location e.g. missingIncludeSystem
-            if (location == null) {
+            if (locations.isEmpty()) {
                 CppcheckNotification.send("no location for " + vFile.getCanonicalPath(),
                         id + " " + severity + " " + inconclusive + " " + errorMessage,
                         NotificationType.ERROR);
                 continue;
             }
 
-            final String fileName = location.file;
-            int lineNumber = location.line;
+            final String fileName = locations.get(0).file;
+            int lineNumber = locations.get(0).line;
             // TODO: use in ProblemDescriptor
-            final int column = location.column;
+            final int column = locations.get(0).column;
 
             if (verboseLevel >= 4) {
                 CppcheckNotification.send(id + " for " + vFile.getCanonicalPath(),
@@ -221,10 +220,29 @@ class CppCheckInspectionImpl {
             final int lineStartOffset = DocumentUtil.getFirstNonSpaceCharOffset(document, lineNumber);
             final int lineEndOffset = document.getLineEndOffset(lineNumber);
 
+            String details = "";
+            if (locations.size() > 1) {
+                final StringBuilder sb = new StringBuilder();
+                for (final Location l : locations) {
+                    if (l.info == null)
+                        continue;
+                    final String name;
+                    if (l.file.equals(sourceFileName)) {
+                        name = vFile.getName(); // replace temporary file name with actual name
+                    } else {
+                        name = l.file;
+                    }
+                    sb.append(String.format("\n%s:%d: note: %s", name, l.line, l.info));
+                }
+                if (sb.length() > 0) {
+                    details = "\n" + sb;
+                }
+            }
+
             final ProblemDescriptor problemDescriptor = manager.createProblemDescriptor(
                     psiFile,
                     TextRange.create(lineStartOffset, lineEndOffset),
-                    "Cppcheck: (" + severity + (inconclusive ? INCONCLUSIVE_TEXT : "") + ") " + id + ": " + errorMessage,
+                    "Cppcheck: (" + severity + (inconclusive ? INCONCLUSIVE_TEXT : "") + ") " + id + ": " + errorMessage + details,
                     severityToHighlightType(severity),
                     true);
             descriptors.add(problemDescriptor);

--- a/src/com/github/johnthagen/cppcheck/CppCheckInspectionImpl.java
+++ b/src/com/github/johnthagen/cppcheck/CppCheckInspectionImpl.java
@@ -62,6 +62,34 @@ class CppCheckInspectionImpl {
     private final int verboseLevel;
     private static final String INCONCLUSIVE_TEXT = ":inconclusive";
 
+    static class Location
+    {
+        Location(@NotNull final Node location) {
+            final NamedNodeMap locationAttributes = location.getAttributes();
+            file = new File(locationAttributes.getNamedItem("file").getNodeValue()).getName();
+            line = Integer.parseInt(locationAttributes.getNamedItem("line").getNodeValue());
+            final Node columnAttr = locationAttributes.getNamedItem("column");
+            // the "column" attribute was added in Cppcheck 1.89
+            if (columnAttr != null) {
+                column = Integer.parseInt(columnAttr.getNodeValue());
+            }
+            else {
+                column = -1;
+            }
+            final Node infoAttr = locationAttributes.getNamedItem("info");
+            if (infoAttr != null) {
+                info = infoAttr.getNodeValue();
+            } else {
+                info = null;
+            }
+        }
+
+        final String file;
+        final int line;
+        final int column;
+        final String info;
+    }
+
     @NotNull
     public List<ProblemDescriptor> parseOutput(@NotNull final PsiFile psiFile,
                                                @NotNull final InspectionManager manager,
@@ -141,14 +169,14 @@ class CppCheckInspectionImpl {
             final Node inconclusiveNode = attributes.getNamedItem("inconclusive");
             final boolean inconclusive = inconclusiveNode != null && inconclusiveNode.getNodeValue().equals("true");
 
-            Node location = null;
+            Location location = null;
 
             // look for the first "location" child name
             final NodeList children = error.getChildNodes();
             for (int j = 0; j < children.getLength(); ++j) {
                 final Node child = children.item(j);
                 if (child.getNodeName().equals("location")) {
-                    location = child;
+                    location = new Location(child);
                     break;
                 }
             }
@@ -161,16 +189,10 @@ class CppCheckInspectionImpl {
                 continue;
             }
 
-            final NamedNodeMap locationAttributes = location.getAttributes();
-            final String fileName = new File(locationAttributes.getNamedItem("file").getNodeValue()).getName();
-            int lineNumber = Integer.parseInt(locationAttributes.getNamedItem("line").getNodeValue());
-            final Node columnAttr = locationAttributes.getNamedItem("column");
+            final String fileName = location.file;
+            int lineNumber = location.line;
             // TODO: use in ProblemDescriptor
-            int column = -1;
-            // the "column" attribute was added in Cppcheck 1.89
-            if (columnAttr != null) {
-                column = Integer.parseInt(columnAttr.getNodeValue());
-            }
+            final int column = location.column;
 
             if (verboseLevel >= 4) {
                 CppcheckNotification.send(id + " for " + vFile.getCanonicalPath(),

